File: 124. Binary Tree Maximum Path Sum/124. Binary Tree Maximum Path Sum.cpp
class Solution {
public:
    int ans = -10000;
    int maxPathSum(TreeNode* root) {
        return max(ans,helper(root));
    }
    int helper(TreeNode* root) {
        if(!root)    return -10000;
        int maxL = helper(root->left);
        int maxR = helper(root->right);
        ans = max({ans, root->val, root->val+maxL+maxR, root->val+max(maxL,maxR)});
        return max(root->val, root->val+max(maxL,maxR));
    }
};

File: 124. Binary Tree Maximum Path Sum/124. Binary Tree Maximum Path Sum.java
class Solution {
    private int maxSum = -1000;
    public int maxPathSum(TreeNode root) {
        return Math.max(subMax(root), maxSum);
    }
    public int subMax(TreeNode curr){
        int leftMax = -1000, rightMax = -1000;
        if(curr.left != null){
            leftMax = subMax(curr.left);
        }
        if(curr.right != null){
            rightMax = subMax(curr.right);
        }
        // int result1 = Stream.of(0, leftMax, rightMax).max(Integer::compare).orElse(0);
        // maxSum = Stream.of(maxSum, curr.val + result1, curr.val + leftMax + rightMax).max(Integer::compare).orElse(0);
        int result1 = Math.max(0, leftMax);
        result1 = Math.max(result1, rightMax);
        maxSum = Math.max(maxSum, curr.val + result1);
        maxSum = Math.max(maxSum, curr.val + leftMax + rightMax);
        return curr.val + result1;
    }
}